src/app/clusters/service-area-server/service-area-server.cpp

```diff
@@ -0,0 +1,1273 @@
+{
+    // give the device a chance to make sure value is up-to-date
+    mDelegate->HandleVolatileCurrentLocation(mCurrentLocation); 
+    return mCurrentLocation;
+}
+
+bool Instance::SetCurrentLocation(const DataModel::Nullable<uint32_t> & aCurrentLocation)
+{
+    // If not null, the value of this attribute SHALL match the LocationID field of an entry on the SupportedLocations attribute's list.
+    if ((!aCurrentLocation.IsNull()) &&
+        (!IsSupportedLocation(aCurrentLocation.Value())))
+    {
+        ChipLogError(Zcl,  "SetCurrentLocation %u - location is not supported", aCurrentLocation.Value());
+        return false;
+    }
+
+    mCurrentLocation = aCurrentLocation;
+
+    // EstimatedEndTime SHALL be null if the CurrentLocation attribute is null.
+    if (mCurrentLocation.IsNull())
+    {
+        SetEstimatedEndTime(DataModel::Nullable<uint32_t>());
+    }
+
+    return true;
+}
+
+
+//*************************************************************************
+// Estimated End Time manipulators
+
+DataModel::Nullable<uint32_t> Instance::GetEstimatedEndTime()
+{
+    // give the device a chance to make sure value is up-to-date
+    mDelegate->HandleVolatileEstimatedEndTime(mEstimatedEndTime);
+    return mEstimatedEndTime;
+}
+
+bool Instance::SetEstimatedEndTime(const DataModel::Nullable<uint32_t> & aEstimatedEndTime)
+{
+    bool ret_value = false;
+    bool doChangeNotify = false;
+
+    // EstimatedEndTime SHALL be null if the CurrentLocation attribute is null.
+    VerifyOrExit((!mCurrentLocation.IsNull() || aEstimatedEndTime.IsNull()),
+        ChipLogError(Zcl,  "SetEstimatedEndTime - must be null if Current Location is null"));
+
+    // The value of this attribute SHALL only be reported in the following cases:
+    // - when it changes from 0 to any other value and vice versa
+    // - when it decreases
+    // - when it changes from null to any non-zero value
+    if ((!mEstimatedEndTime.IsNull() && !aEstimatedEndTime.IsNull() && (aEstimatedEndTime.Value() == 0) && mEstimatedEndTime.Value()!= 0)   ||
+        (!mEstimatedEndTime.IsNull() && !aEstimatedEndTime.IsNull() && (aEstimatedEndTime.Value() != 0) && mEstimatedEndTime.Value()== 0)   ||
+        (!mEstimatedEndTime.IsNull() && !aEstimatedEndTime.IsNull() && (aEstimatedEndTime.Value() < mEstimatedEndTime.Value()))             ||
+        (mEstimatedEndTime.IsNull()  && !aEstimatedEndTime.IsNull() && (aEstimatedEndTime.Value() != 0))  )
+    {
+        doChangeNotify = true;
+    }
+
+    mEstimatedEndTime = aEstimatedEndTime;
+
+    if (doChangeNotify)
+    {
+        NotifyEstimatedEndTimeChanged();
+    }
+
+    // success
+    ret_value = true;
+
+
+exit:
+    return ret_value;
+}
+
+
+//*************************************************************************
+// Progress list manipulators
+
+bool Instance::AddPendingProgressElement(uint32_t aLocationId)
+{
+    bool     ret_value = false;
+    uint32_t dummyIndex;
```

Comment:
Does matter use the convention of having and `_` at the start of a variable that is intended not to be used?

```suggestion
    uint32_t _dummyIndex;
```